File: pi_scripts/pi-fleet-manager/frontend/js/commands.js
const API_BASE = '/api';

let hosts = [];
let keys = [];
let currentStreamConnection = null;

async function loadData() {
    try {
        const [hostsRes, keysRes] = await Promise.all([
            fetch(`${API_BASE}/hosts/`),
            fetch(`${API_BASE}/keys/`)
        ]);
        
        const hostsData = await hostsRes.json();
        const keysData = await keysRes.json();
        
        hosts = hostsData.hosts;
        keys = keysData.keys;
        
        populateHostCheckboxes('commandHosts');
        populateKeySelect('commandKey');
    } catch (error) {
        console.error('Error loading data:', error);
        showError('Failed to load hosts and keys');
    }
}

function populateHostCheckboxes(containerId) {
    const container = document.getElementById(containerId);
    container.innerHTML = hosts.map(host => `
        <label class="checkbox-label">
            <input type="checkbox" value="${host.id}">
            ${escapeHtml(host.name)} (${escapeHtml(host.host)})
        </label>
    `).join('');
}

function populateKeySelect(selectId) {
    const select = document.getElementById(selectId);
    select.innerHTML = keys.map(key => `
        <option value="${key.id}">${escapeHtml(key.name)}</option>
    `).join('');
}

function getSelectedHosts() {
    const checkboxes = document.querySelectorAll('#commandHosts input[type="checkbox"]:checked');
    return Array.from(checkboxes).map(cb => cb.value);
}

async function executeCommand() {
    const hostIds = getSelectedHosts();
    const username = document.getElementById('commandUsername').value;
    const keyId = document.getElementById('commandKey').value;
    const command = document.getElementById('commandInput').value;

    if (hostIds.length === 0) {
        showError('Please select at least one host');
        return;
    }

    if (!username || !command) {
        showError('Please fill in all required fields');
        return;
    }

    const commandData = {
        command: command,
        host_ids: hostIds,
        username: username,
        key_id: keyId
    };

    try {
        const res = await fetch(`${API_BASE}/commands/execute`, {
            method: 'POST',
            headers: {'Content-Type': 'application/json'},
            body: JSON.stringify(commandData)
        });
        
        const results = await res.json();
        displayCommandResults(results.results);
    } catch (error) {
        console.error('Error executing command:', error);
        showError('Failed to execute command');
    }
}

function executeCommandStream() {
    const hostIds = getSelectedHosts();
    const username = document.getElementById('commandUsername').value;
    const keyId = document.getElementById('commandKey').value;
    const command = document.getElementById('commandInput').value;

    if (hostIds.length === 0) {
        showError('Please select at least one host');
        return;
    }

    if (!username || !command) {
        showError('Please fill in all required fields');
        return;
    }

    // Create WebSocket connection for streaming
    const connectionId = 'cmd_' + Date.now();
    currentStreamConnection = new WebSocket(`ws://${window.location.host}/api/commands/stream/${connectionId}`);
    
    // Clear previous results
    const resultsContainer = document.getElementById('command-results');
    resultsContainer.innerHTML = '';
    
    // Create output boxes for each host
    hostIds.forEach(hostId => {
        const host = hosts.find(h => h.id === hostId);
        if (host) {
            const outputBox = document.createElement('div');
            outputBox.className = 'output-box';
            outputBox.id = `output-${hostId}`;
            outputBox.innerHTML = `
                <div class="output-box-header">${escapeHtml(host.name)}</div>
                <pre></pre>
            `;
            resultsContainer.appendChild(outputBox);
        }
    });

    currentStreamConnection.onopen = () => {
        // Send command data
        currentStreamConnection.send(JSON.stringify({
            command: command,
            host_ids: hostIds,
            username: username,
            key_id: keyId
        }));
    };

    currentStreamConnection.onmessage = (event) => {
        const data = JSON.parse(event.data);
        
        if (data.host_id) {
            const outputBox = document.getElementById(`output-${data.host_id}`);
            if (outputBox) {
                const pre = outputBox.querySelector('pre');
                
                if (data.status === 'connecting') {
                    pre.textContent += `\n🔗 Connecting to ${data.host_name}...\n`;
                } else if (data.status === 'executing') {
                    pre.textContent += `🚀 Executing command...\n`;
                } else if (data.status === 'complete') {
                    pre.textContent += `✅ Command completed (exit code: ${data.exit_code})\n`;
                    if (data.stdout) pre.textContent += `\nSTDOUT:\n${data.stdout}\n`;
                    if (data.stderr) pre.textContent += `\nSTDERR:\n${data.stderr}\n`;
                } else if (data.status === 'error') {
                    pre.textContent += `❌ Error: ${data.error}\n`;
                }
                
                outputBox.scrollTop = outputBox.scrollHeight;
            }
        }
        
        if (data.status === 'all_complete') {
            showSuccess('All commands completed');
            currentStreamConnection.close();
            currentStreamConnection = null;
        }
    };

    currentStreamConnection.onerror = (error) => {
        console.error('WebSocket error:', error);
        showError('WebSocket connection failed');
        currentStreamConnection = null;
    };
}

function displayCommandResults(results) {
    const container = document.getElementById('command-results');
    container.innerHTML = '';

    for (const [hostId, result] of Object.entries(results)) {
        const host = hosts.find(h => h.id === hostId) || { name: hostId };
        
        const outputBox = document.createElement('div');
        outputBox.className = 'output-box';
        
        let content = `Host: ${escapeHtml(result.host_name || host.name)}\n`;
        content += `Address: ${escapeHtml(result.host || 'Unknown')}\n\n`;
        
        if (result.error) {
            content += `❌ ERROR: ${escapeHtml(result.error)}\n`;
        } else {
            if (result.stdout) {
                content += `STDOUT:\n${escapeHtml(result.stdout)}\n`;
            }
            if (result.stderr) {
                content += `STDERR:\n${escapeHtml(result.stderr)}\n`;
            }
            content += `\nExit Code: ${result.exit_code}\n`;
        }
        
        outputBox.innerHTML = `
            <div class="output-box-header">${escapeHtml(result.host_name || host.name)}</div>
            <pre>${content}</pre>
        `;
        
        container.appendChild(outputBox);
    }
}

function escapeHtml(text) {
    const div = document.createElement('div');
    div.textContent = text;
    return div.innerHTML;
}

function showSuccess(message) {
    const container = document.querySelector('.container');
    const alert = document.createElement('div');
    alert.className = 'alert alert-success';
    alert.textContent = message;
    container.insertBefore(alert, container.firstChild);
    
    setTimeout(() => alert.remove(), 5000);
}

function showError(message) {
    const container = document.querySelector('.container');
    const alert = document.createElement('div');
    alert.className = 'alert alert-danger';
    alert.textContent = message;
    container.insertBefore(alert, container.firstChild);
    
    setTimeout(() => alert.remove(), 5000);
}

// Load data on page load
document.addEventListener('DOMContentLoaded', loadData);

File: pi-fleet-manager/api/commands.py
from fastapi import APIRouter, WebSocket, WebSocketDisconnect, HTTPException
from pydantic import BaseModel
from typing import List, Dict
import asyncio
import json
from core.ssh_manager import SSHManager
from core.database import get_item, get_collection

router = APIRouter()
ssh_manager = SSHManager()

class CommandExecute(BaseModel):
    command: str
    host_ids: List[str]
    username: str
    key_id: str

@router.post("/execute")
async def execute_command(cmd_data: CommandExecute):
    """Execute command on multiple hosts"""
    # Get key
    key = get_item("ssh_keys", cmd_data.key_id)
    if not key:
        raise HTTPException(status_code=404, detail="Key not found")
    
    results = {}
    
    for host_id in cmd_data.host_ids:
        host = get_item("hosts", host_id)
        if not host:
            results[host_id] = {
                "error": "Host not found",
                "host_name": host_id
            }
            continue
        
        try:
            # Get jumphost if needed
            jumphost = None
            if host.get('jumphost_id'):
                jumphost_data = get_item("hosts", host['jumphost_id'])
                if jumphost_data:
                    jumphost = {
                        'host': jumphost_data['host'],
                        'port': jumphost_data.get('port', 22),
                        'username': cmd_data.username
                    }
            
            # Connect and execute
            client = ssh_manager.connect(
                host['host'],
                host.get('port', 22),
                cmd_data.username,
                key['private_key'],
                jumphost
            )
            
            result = ssh_manager.execute_command(client, cmd_data.command)
            ssh_manager.close(client)
            
            results[host_id] = {
                "host_name": host['name'],
                "host": host['host'],
                "stdout": result['stdout'],
                "stderr": result['stderr'],
                "exit_code": result['exit_code']
            }
            
        except Exception as e:
            results[host_id] = {
                "host_name": host.get('name', host_id),
                "host": host.get('host', ''),
                "error": str(e),
                "exit_code": 1
            }
    
    return {"results": results}

class WebSocketManager:
    """Manage WebSocket connections for streaming command output"""
    
    def __init__(self):
        self.active_connections: Dict[str, WebSocket] = {}
    
    async def connect(self, websocket: WebSocket, connection_id: str):
        await websocket.accept()
        self.active_connections[connection_id] = websocket
    
    def disconnect(self, connection_id: str):
        if connection_id in self.active_connections:
            del self.active_connections[connection_id]
    
    async def send_message(self, connection_id: str, message: dict):
        if connection_id in self.active_connections:
            try:
                await self.active_connections[connection_id].send_json(message)
            except:
                self.disconnect(connection_id)

ws_manager = WebSocketManager()

@router.websocket("/stream/{connection_id}")
async def websocket_endpoint(websocket: WebSocket, connection_id: str):
    """WebSocket endpoint for streaming command execution"""
    await ws_manager.connect(websocket, connection_id)
    
    try:
        while True:
            data = await websocket.receive_json()
            
            # Expected data format:
            # {
            #   "command": "ls -la",
            #   "host_ids": ["host1", "host2"],
            #   "username": "pi",
            #   "key_id": "key123"
            # }
            
            command = data.get('command')
            host_ids = data.get('host_ids', [])
            username = data.get('username')
            key_id = data.get('key_id')
            
            if not all([command, host_ids, username, key_id]):
                await ws_manager.send_message(connection_id, {
                    "error": "Missing required fields"
                })
                continue
            
            # Get key
            key = get_item("ssh_keys", key_id)
            if not key:
                await ws_manager.send_message(connection_id, {
                    "error": "Key not found"
                })
                continue
            
            # Execute on each host
            for host_id in host_ids:
                host = get_item("hosts", host_id)
                if not host:
                    await ws_manager.send_message(connection_id, {
                        "host_id": host_id,
                        "error": "Host not found"
                    })
                    continue
                
                try:
                    # Get jumphost if needed
                    jumphost = None
                    if host.get('jumphost_id'):
                        jumphost_data = get_item("hosts", host['jumphost_id'])
                        if jumphost_data:
                            jumphost = {
                                'host': jumphost_data['host'],
                                'port': jumphost_data.get('port', 22),
                                'username': username
                            }
                    
                    # Send start message
                    await ws_manager.send_message(connection_id, {
                        "host_id": host_id,
                        "host_name": host['name'],
                        "status": "connecting"
                    })
                    
                    # Connect
                    client = ssh_manager.connect(
                        host['host'],
                        host.get('port', 22),
                        username,
                        key['private_key'],
                        jumphost
                    )
                    
                    await ws_manager.send_message(connection_id, {
                        "host_id": host_id,
                        "host_name": host['name'],
                        "status": "executing"
                    })
                    
                    # Execute command
                    result = ssh_manager.execute_command(client, command)
                    ssh_manager.close(client)
                    
                    # Send result
                    await ws_manager.send_message(connection_id, {
                        "host_id": host_id,
                        "host_name": host['name'],
                        "status": "complete",
                        "stdout": result['stdout'],
                        "stderr": result['stderr'],
                        "exit_code": result['exit_code']
                    })
                    
                except Exception as e:
                    await ws_manager.send_message(connection_id, {
                        "host_id": host_id,
                        "host_name": host.get('name', host_id),
                        "status": "error",
                        "error": str(e)
                    })
            
            # Send completion message
            await ws_manager.send_message(connection_id, {
                "status": "all_complete"
            })
            
    except WebSocketDisconnect:
        ws_manager.disconnect(connection_id)
    except Exception as e:
        print(f"WebSocket error: {e}")
        ws_manager.disconnect(connection_id)